lib/SILGen/SILGenStmt.cpp

```diff
@@ -379,18 +379,20 @@ prepareIndirectResultInit(SILGenFunction &SGF,
                           SmallVectorImpl<CleanupHandle> &cleanups) {
   // Recursively decompose tuple types.
   if (auto resultTupleType = dyn_cast<TupleType>(resultType)) {
-    auto tupleInit = new TupleInitialization();
-    tupleInit->SubInitializations.reserve(resultTupleType->getNumElements());
-
-    for (auto resultEltType : resultTupleType.getElementTypes()) {
-      auto eltInit = prepareIndirectResultInit(SGF, fnTypeForResults,
-                                               resultEltType, allResults,
-                                               directResults,
-                                               indirectResultAddrs, cleanups);
-      tupleInit->SubInitializations.push_back(std::move(eltInit));
-    }
+    // If this is an empty tuple, we don't want to decompose it.
+    if (resultTupleType.getElementTypes().size()) {
```

Comment:
Did this need to be changed?

Reply:
IOW, I don't understand why this change requires this other fix. Does a crash result? Please explain this to me?

Reply:
Yes. Using any type with a user-provided destructor crashes without this fix. 

When we have an empty tuple, we get the wrong return type later on. Because the value is returned with an indirect result, the return type should be `@out Foo` but instead is `()` (because that's what's returned).

Reply:
TBH, I'm not sure this is the right fix. It did fix the problem and (seems like?) it doesn't have any unwanted side effects but, it doesn't fix the "root" of the issue. I can investigate this a bit more.

Reply:
Yes. Maybe @jckarter or @slavapestov have ideas.

Reply:
Yeah, it seems like a reasonable optimization to avoid forming a tuple initialization for an empty tuple, but it doesn't seem like it should be necessary for correctness. How does it crash without this change?

Reply:
@jckarter the assertion on line 451: `assert(allResults.empty());`. 

Because in `prepareIndirectResultInit` the tuple path is taken and the tuple is empty, the results aren't recursively `slice`d off `allResults`.

Reply:
FWIW I suspect this is probably slower (it goes from essentially a no-op to a recursive function that _will_ dynamically allocate memory).

Reply:
@zoecarver If the result type is an empty tuple, though, then there shouldn't be anything in `allResults` to slice off to begin with. Do you know what the root of the change in behavior is?

Reply:
@jckarter sorry this comment slipped through my inbox somehow. I can spend some time today looking into this in more depth. 

> Do you know what the root of the change in behavior is?

I don't know.

> If the result type is an empty tuple, though, then there shouldn't be anything in allResults to slice off to begin with.

I think `allResults` holds indirect results (also?).

Reply:
I was able to find and fix the root of the problem (the return statement was given an expression when it shouldn't have).